File: p008.py
from functools import reduce # Valid in Python 2.6+, required in Python 3, from https://stackoverflow.com/questions/595374/whats-the-function-like-sum-but-for-multiplication-product
import operator

def get_greatest_product_list(digits, number_of_adjacent_digits):
    greatest_product = max(
        reduce(operator.mul, digits[i:i+number_of_adjacent_digits], 1)
        for i in range(len(digits) - number_of_adjacent_digits + 1))
    return greatest_product

def test_get_greatest_product_list_at_start():
    digits = list(reversed(range(4)))
    result = get_greatest_product_list(digits, 2)
    assert result == 3 * 2

def test_get_greatest_product_list_at_end():
    digits = list(range(4))
    result = get_greatest_product_list(digits, 2)
    assert result == 2 * 3

def get_digits_from_file():
    with open("p008.input") as input_file:
        digits = list(int(c) for c in input_file.read().replace("\n", ""))
    return digits

def test_get_digits_from_file():
    digits = get_digits_from_file()
    assert len(digits) == 1000 # problem description: file contains 1000 digit number

def get_greatest_product_file(number_of_adjacent_digits):
    digits = get_digits_from_file()
    return get_greatest_product_list(digits, number_of_adjacent_digits)

def test_example():
    assert get_greatest_product_file(4) == 5832

if __name__ == "__main__":
    solution = get_greatest_product_file(13)
    print(solution)

File: p008.rs
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::Path;

fn get_max_product_vector(digits: Vec<u8>, length: usize) -> u64 {
    digits.windows(length)
        .map(|window| window.iter().map(|&d| d as u64).product())
        .max()
        .unwrap()
}

#[test]
fn test_get_max_product_vector_at_end() {
    let digits = (0..4).into_iter().collect();
    let result = get_max_product_vector(digits, 2);
    assert_eq!(result, 2 * 3);
}

#[test]
fn test_get_max_product_vector_at_start() {
    let digits = (0..4).into_iter().rev().collect();
    let result = get_max_product_vector(digits, 2);
    assert_eq!(result, 3 * 2);
}

fn get_digits_from_file() -> Vec<u8> {
    let path = Path::new("p008.input");
    let file = File::open(&path).expect("Failed to open file");
    let mut reader = BufReader::new(file);
    let mut contents = String::new();
    reader.read_to_string(&mut contents).expect("Failed to read file");

    let digits: Vec<u8> = contents
        .chars()
        .filter(|c| c.is_digit(10))
        .map(|c| c.to_digit(10).unwrap() as u8)
        .collect();

    digits
}

#[test]
fn test_get_digits_from_file() {
    let digits = get_digits_from_file();
    assert_eq!(digits.len(), 1000); // problem description: file contains 1000 digit number
}

fn get_max_product_file(length: usize) -> u64 {
    let digits = get_digits_from_file();
    let max_product = get_max_product_vector(digits, length);
    max_product
}

#[test]
fn test_get_max_product_file_example() {
    let result = get_max_product_file(4);
    assert_eq!(result, 5832);
}

fn main() {
    let solution = get_max_product_file(13);
    println!("{}", solution);
}